File: frontend/src/api/diff.ts
import { useQuery } from "@tanstack/react-query";
import apiClient from "./client";

export interface DiffItem {
  rule_id: string;
  title: string;
  status: string;
  change_type: string;
  changed_at: string;
}

export interface PaginatedDiff {
  items: DiffItem[];
  total: number;
  page: number;
  limit: number;
}

export interface DiffVersion {
  definition: string | null;
  status: string | null;
  changed_at: string;
  change_note: string | null;
}

export interface DiffDetail {
  rule_id: string;
  rule_title: string;
  before: DiffVersion | null;
  after: DiffVersion | null;
  versions: DiffVersion[];
}

export const useDiffList = (page = 1) =>
  useQuery({
    queryKey: ["diff", page],
    queryFn: async () => {
      const res = await apiClient.get<PaginatedDiff>("/diff", { params: { page } });
      return res.data;
    },
  });

export const useRuleDiff = (ruleId: string) =>
  useQuery({
    queryKey: ["diff", ruleId],
    queryFn: async () => {
      const res = await apiClient.get<DiffDetail>(`/diff/${ruleId}`);
      return res.data;
    },
    enabled: !!ruleId,
  });

File: app/schemas/diff.py
"""Pydantic schemas for diff endpoints."""
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class DiffItem(BaseModel):
    rule_id: uuid.UUID
    title: str
    status: str
    change_type: str
    changed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class PaginatedDiff(BaseModel):
    items: list[DiffItem]
    total: int
    page: int
    limit: int


class DiffVersion(BaseModel):
    definition: str
    status: Optional[str] = None
    changed_at: Optional[datetime] = None
    change_note: Optional[str] = None


class DiffDetail(BaseModel):
    rule_id: uuid.UUID
    rule_title: str
    before: Optional[DiffVersion] = None
    after: Optional[DiffVersion] = None
    versions: list[DiffVersion] = []
